### PR TITLE
Do not initialize the same processor instance multiple times

### PR DESCRIPTION
**Pull request by hunterhector**
This PR fixes https://github.com/asyml/forte/issues/410. 

### Description of changes
1. Add a flag for each `PipelineComponent` marking the initialization status.
2. When the pipeline is initializing, it will check whether the flag is True. It will skip initializing the processor if the flag is True.

### Possible influences of this PR.
Describe what are the possible side-effects of the code change.

### Test Conducted
A reuse-processor test is added. The testing processor will record the number of times it gets initialized. And the number should be 1.

### CODE REVIEW

File: forte/pipeline.py [hunk 1]
```diff
@@ -184,55 +187,173 @@ def init_from_config(self, configs: List):
...
             except ProcessorConfigError as e:
                 logging.error("Exception occur when initializing "
-                              "processor %s", processor.name)
+                              "processor %s", component.name)
                 raise e
 
+            if self._check_type_consistency:
+                component.enforce_consistency(enforce=True)
+            else:
+                component.enforce_consistency(enforce=False)
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/pipeline.py:262]: yeah, this is from another PR, but I can change this too.

File: forte/pipeline_component.py [hunk 2]
```diff
@@ -106,14 +135,14 @@ def finish(self, resource: Resources):
         Args:
             resource (Resources): A global resource registry.
         """
-        pass
+        self.__is_initialized = False
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/pipeline_component.py:138]: Yes, if you called `run` which runs over the whole life-cycle (initialize, process, finish). You can often call initialize and process manually, without calling finish.

The `finish` stage is designed to release all resources.  But I also realize we are missing one stage that does not release the resources but collect some aggregated information. But maybe that belongs to another PR.

File: forte/pipeline_component.py [hunk 3]
```diff
@@ -72,6 +88,18 @@ def initialize(self, resources: Resources, configs: Config):
         """
         self.resources = resources
         self.configs = configs
+        self.__is_initialized = True
+
+    def reset_flag(self):
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/pipeline_component.py:93]: renamed.

File: forte/pipeline_component.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/pipeline_component.py:138]: yes if we run pipeline via `run`, which will go through the whole lifecycle of the pipeline. We can also do `initialize`, `process_dataset`, but do not release the resources here. The `finish` call is designed to release the resources.

File: tests/forte/pipeline_test.py [hunk 4]
```diff
@@ -1046,6 +1057,69 @@ def test_pipeline5(self):
...
+            MultiPackBoxer()).add(
+            MultiPackCopier())
+
+        # Create one shared instance of this extractor
+        dummy = DummyPackProcessor()
+        nlp.add(dummy, config={"test": "dummy1"},
+                selector=NameMatchSelector("default"))
+        # This will not change config because the processor is initialized.
+        #  but the selector should work.
+        nlp.add(dummy, config={"test": "dummy2"},
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply tests/forte/pipeline_test.py:1073]: Added a check.

File: forte/pipeline.py [hunk 5]
```diff
@@ -184,55 +187,173 @@ def init_from_config(self, configs: List):
...
+        initialized (for example, being initialized manually or used twice
+        in the same pipeline), the new configuration will be ignored.
+
+        The pipeline will check for type dependencies between the components
+        inside this pipeline, see
+        :func:`~forte.pipeline_component.PipelineComponent.enforce_consistency`
+        for more details.
+
+        """
+        for component, config in zip(self.components, self.component_configs):
```
[the remark this answers is not in the document]
Comment (hunterhector) [reply forte/pipeline.py:246]: yeah. So I haven't come up with the good name. Now there are different types of components, where `reader` is handled separately. `processor`, `caster`, `evaluator` are the remaining component types. I will see if I can rename the function too.

[the remark this answers is not in the document]
Comment (DianLiI) [reply forte/pipeline.py:253]: Not a big deal, I think it can be deferred to another PR.

**Review by DianLiI — DISMISSED** [review]

**Pull request merged**